frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/base/link_to_page_widget.dart

```diff
@@ -92,85 +93,167 @@ class LinkToPageMenu extends StatefulWidget {
-          boxShadow: [
-            BoxShadow(
-              blurRadius: 5,
-              spreadRadius: 1,
-              color: Colors.black.withOpacity(0.1),
-            ),
-          ],
-          borderRadius: BorderRadius.circular(6.0),
+        color: Colors.transparent,
+        width: 300,
+        child: Container(
+          padding: const EdgeInsets.fromLTRB(10, 6, 10, 6),
+          decoration: BoxDecoration(
+            color: style.selectionMenuBackgroundColor,
+            boxShadow: [
+              BoxShadow(
+                blurRadius: 5,
+                spreadRadius: 1,
+                color: Colors.black.withOpacity(0.1),
+              ),
+            ],
+            borderRadius: BorderRadius.circular(6.0),
+          ),
+          child: _isLoading
+              ? const Center(
+                  child: CircularProgressIndicator(),
+                )
+              : _buildListWidget(context, _selectedIndex, _availableLayout),
         ),
-        child: _buildListWidget(context),
       ),
     );
   }
 
-  Widget _buildListWidget(BuildContext context) {
-    return FutureBuilder<List<dartz.Tuple2<AppPB, List<ViewPB>>>>(
-      builder: (context, snapshot) {
-        if (snapshot.hasData &&
-            snapshot.connectionState == ConnectionState.done) {
-          final apps = snapshot.data;
-          final children = <Widget>[
-            Padding(
-              padding: const EdgeInsets.symmetric(vertical: 4),
-              child: FlowyText.regular(
-                widget.hintText,
-                fontSize: 10,
-                color: Colors.grey,
+  KeyEventResult _onKey(FocusNode node, RawKeyEvent event) {
+    if (event is! RawKeyDownEvent || _isLoading || _availableLayout.isEmpty) {
+      return KeyEventResult.ignored;
+    }
+
+    final acceptedKeys = [
+      LogicalKeyboardKey.arrowUp,
+      LogicalKeyboardKey.arrowDown,
+      LogicalKeyboardKey.tab,
+      LogicalKeyboardKey.enter
+    ];
+
+    if (!acceptedKeys.contains(event.logicalKey)) {
+      return KeyEventResult.handled;
+    }
+
+    var newSelectedIndex = _selectedIndex;
+    if (event.logicalKey == LogicalKeyboardKey.arrowDown &&
+        newSelectedIndex != _totalItems - 1) {
+      newSelectedIndex += 1;
+    } else if (event.logicalKey == LogicalKeyboardKey.arrowUp &&
+        newSelectedIndex != 0) {
+      newSelectedIndex -= 1;
+    } else if (event.logicalKey == LogicalKeyboardKey.tab) {
+      newSelectedIndex += 1;
+      newSelectedIndex %= _totalItems;
+    } else if (event.logicalKey == LogicalKeyboardKey.enter) {
+      widget.onSelected(
+          _items[_selectedIndex]!.value1, _items[_selectedIndex]!.value2);
+    }
+
+    setState(() {
+      _selectedIndex = newSelectedIndex;
+    });
+
+    return KeyEventResult.handled;
+  }
+
+  Widget _buildListWidget(BuildContext context, int selectedIndex,
+      List<dartz.Tuple2<AppPB, List<ViewPB>>> items) {
+    int index = 0;
+    return Builder(builder: (context) {
```

Comment:
Removed the FutureBuilder and replaced it with a Builder whose elements depend on a Future (`fetchAll`). The parts of `initState` that request focus are appropriate, but I'd avoid running async functions in initState. You cannot await async functions in initState by design. So I'd like this code to be refactored such that:
1. `fetchAll` function returns the data you need in this widget.
2. Make this widget a FutureBuilder.

Reply:
Hey actually the reason I shifted from FutureBuilder to Builder is that with the former with every index change, it was rebuilding the widget (fetching again the items), so to avoid that I decided to fetch items separately at the build time. I will make the required changes and will again attach the video, and let me know if its good to go. Thank you!

Reply:
Ah I see, that makes sense. If you make another widget to track the index, and return that widget from the Future builder, I believe only the widget that tracks the index will rebuild when the index is updated, and the data won't be fetched every time. Thanks for the clarification. Will keep my eyes on this PR, thank you.

Reply:
Hey so I am not quite sure how to go about it. Since whenever _selectedIndex gets updated the build method would trigger again. And since _buildListWidget is part of it, it will trigger every time there is setState called. Using initState the call to fetch data is made only once.

I also searched over internet, and found this article helpful : https://medium.com/flutterworld/why-future-builder-called-multiple-times-9efeeaf38ba2. And I suppose I am using the first method mentioned in the article.

So how to go about it, is there something that I am doing wrong?

Reply:
To avoid building it every time, you can create the future during the initState.  https://api.flutter.dev/flutter/widgets/FutureBuilder-class.html